virtual/activation_cache.go

```diff
@@ -102,18 +109,31 @@ func (a *activationsCache) ensureActivation(
 	bufIface, cacheKey = actorCacheKeyUnsafePooled(namespace, moduleID, actorID)
 	aceI, ok := a.c.Get(cacheKey)
 	bufPool.Put(bufIface)
-	// Cache miss, fill the cache.
-	if !ok ||
+
+	hasBlacklistedID := false
```

Comment:
This block of code needs a comment IMO. I *think* the logic is something like:

"If the request thinks some servers should be blacklisted, and any of the servers the current request thinks should be blacklisted are not marked as blacklisted in the cache, then we should force the cache to be refreshed and ignore the existing entry, otherwise it might end up routing us back to the blacklisted server ID."

The next obvious question to me then though is if we have multiple replicas and one server is blacklisted, but the other is not, why not allow the non-blacklisted ID to be used from the cache and let the cache refresh itself eventually in the background automatically? The only issue with that is the caller will only know about one blacklisted server ID at a time whereas the cache needs to remember all of them. This could be solved though by remembering the accumulated set of blacklisted server IDs in the cache and then including that entire set in the refresh request.

So then the way this would work is everytime the caller comes with a blacklisted server ID you do something like the following:

1. Add the current blacklisted server ID to a set of blacklisted server IDs in the cache (I.E the cache values need to updated/mutated)
3. Check if all the server IDs in the cache are blacklisted. If so, force a cache update immediately (there is nothing else to do really)
4. Otherwise, just return the subset of server IDs that are not blacklisted yet

The reason I care about this is when the registry is down (for whatever reason) we don't want to evict the cache entries if some of the values are still safe to use. Although I guess balancing only runs when the registry is not down so maybe its not a huge deal to do it the way you have it currently 🤔 

Something to think about for sure though. I wonder if a lot of this logic would be simplified if the server that blacklists the actor just notified the registry of that via a dedicated method instead of forcing all the clients to do it on its behalf